File: examples/php/health_endpoint.php
<?php

/**
 * Health Endpoint Example for PHP
 *
 * This demonstrates how to implement health check endpoints that work with
 * the Status Incident Service's heartbeat monitoring feature.
 *
 * For standalone PHP:
 *   php -S localhost:8080 health_endpoint.php
 *
 * Then configure Status Incident to monitor: http://localhost:8080/health
 */

declare(strict_types=1);

/**
 * Health check response builder.
 */
class HealthChecker
{
    private ?PDO $db = null;
    private ?Redis $redis = null;

    public function __construct(?PDO $db = null, ?Redis $redis = null)
    {
        $this->db = $db;
        $this->redis = $redis;
    }

    /**
     * Perform all health checks.
     */
    public function check(): array
    {
        $checks = [];
        $healthy = true;

        // Database check
        if ($this->db !== null) {
            try {
                $this->db->query('SELECT 1');
                $checks['database'] = 'ok';
            } catch (PDOException $e) {
                $checks['database'] = 'error: ' . $e->getMessage();
                $healthy = false;
            }
        } else {
            $checks['database'] = 'ok (mock)';
        }

        // Redis check
        if ($this->redis !== null) {
            try {
                $this->redis->ping();
                $checks['redis'] = 'ok';
            } catch (RedisException $e) {
                $checks['redis'] = 'error: ' . $e->getMessage();
                $healthy = false;
            }
        } else {
            $checks['redis'] = 'ok (mock)';
        }

        return [
            'status' => $healthy ? 'ok' : 'error',
            'checks' => $checks,
            'healthy' => $healthy,
        ];
    }
}

/**
 * Simple router for the health endpoints.
 */
function handleRequest(): void
{
    $uri = parse_url($_SERVER['REQUEST_URI'], PHP_URL_PATH);
    
    // Initialize health checker (replace with your actual connections)
    $checker = new HealthChecker();

    header('Content-Type: application/json');

    switch ($uri) {
        case '/health':
            $result = $checker->check();
            if (!$result['healthy']) {
                http_response_code(503);
            }
            unset($result['healthy']);
            echo json_encode($result);
            break;

        case '/health/live':
            // Kubernetes liveness probe - is the application running?
            echo json_encode(['status' => 'ok']);
            break;

        case '/health/ready':
            // Kubernetes readiness probe - is the application ready?
            $result = $checker->check();
            if (!$result['healthy']) {
                http_response_code(503);
            }
            echo json_encode(['status' => $result['status']]);
            break;

        default:
            http_response_code(404);
            echo json_encode(['error' => 'Not found']);
    }
}

// Run the router
handleRequest();

File: examples/go/health_endpoint.go
// Health endpoint example for Go applications.
//
// This demonstrates how to implement health check endpoints that work with
// the Status Incident Service's heartbeat monitoring feature.
//
// Run with:
//
//	go run health_endpoint.go
//
// Then configure Status Incident to monitor: http://localhost:8080/health
package main

import (
	"context"
	"database/sql"
	"encoding/json"
	"log"
	"net/http"
	"time"
)

// HealthResponse represents the health check response.
type HealthResponse struct {
	Status  string            `json:"status"`
	Checks  map[string]string `json:"checks,omitempty"`
	Message string            `json:"message,omitempty"`
}

// HealthChecker performs health checks on dependencies.
type HealthChecker struct {
	db    *sql.DB
	redis interface{ Ping(context.Context) error }
}

// NewHealthChecker creates a new health checker.
// Replace with your actual database and cache connections.
func NewHealthChecker() *HealthChecker {
	return &HealthChecker{
		db:    nil, // Replace with: sql.Open("postgres", connStr)
		redis: nil, // Replace with: redis.NewClient(&redis.Options{...})
	}
}

// Check performs all health checks.
func (h *HealthChecker) Check(ctx context.Context) HealthResponse {
	response := HealthResponse{
		Status: "ok",
		Checks: make(map[string]string),
	}
	healthy := true

	// Database check
	if h.db != nil {
		if err := h.db.PingContext(ctx); err != nil {
			response.Checks["database"] = "error: " + err.Error()
			healthy = false
		} else {
			response.Checks["database"] = "ok"
		}
	} else {
		response.Checks["database"] = "ok (mock)"
	}

	// Redis check
	if h.redis != nil {
		if err := h.redis.Ping(ctx); err != nil {
			response.Checks["redis"] = "error: " + err.Error()
			healthy = false
		} else {
			response.Checks["redis"] = "ok"
		}
	} else {
		response.Checks["redis"] = "ok (mock)"
	}

	if !healthy {
		response.Status = "error"
		response.Message = "One or more dependencies are unhealthy"
	}

	return response
}

// HealthHandler handles /health requests.
func HealthHandler(checker *HealthChecker) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		ctx, cancel := context.WithTimeout(r.Context(), 5*time.Second)
		defer cancel()

		response := checker.Check(ctx)

		w.Header().Set("Content-Type", "application/json")
		if response.Status != "ok" {
			w.WriteHeader(http.StatusServiceUnavailable)
		}
		json.NewEncoder(w).Encode(response)
	}
}

// LivenessHandler handles /health/live requests (Kubernetes liveness probe).
func LivenessHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]string{"status": "ok"})
}

// ReadinessHandler handles /health/ready requests (Kubernetes readiness probe).
func ReadinessHandler(checker *HealthChecker) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		ctx, cancel := context.WithTimeout(r.Context(), 5*time.Second)
		defer cancel()

		response := checker.Check(ctx)

		w.Header().Set("Content-Type", "application/json")
		if response.Status != "ok" {
			w.WriteHeader(http.StatusServiceUnavailable)
		}
		json.NewEncoder(w).Encode(map[string]string{"status": response.Status})
	}
}

func main() {
	checker := NewHealthChecker()

	mux := http.NewServeMux()
	mux.HandleFunc("/health", HealthHandler(checker))
	mux.HandleFunc("/health/live", LivenessHandler)
	mux.HandleFunc("/health/ready", ReadinessHandler(checker))

	log.Println("Starting health endpoint example on http://localhost:8080")
	log.Println("Endpoints:")
	log.Println("  - GET /health       - Full health check")
	log.Println("  - GET /health/live  - Liveness probe")
	log.Println("  - GET /health/ready - Readiness probe")

	if err := http.ListenAndServe(":8080", mux); err != nil {
		log.Fatal(err)
	}
}
